File: MinJu/240101_week/Programmers_154540_무인도 여행.py
from collections import deque

def solution(maps):
    answer = []
    # 범위
    n = len(maps)  # 세로
    m = len(maps[0])  # 가로
    visited = [[0] * m for _ in range(n)]  # 방문 여부

    # 상하좌우 방향
    d = [(0, 1), (0, -1), (1, 0), (-1, 0)]

    for i in range(n):
        for j in range(m):
            # 바다이거나 이미 방문한 곳이면 무시
            if maps[i][j] == 'X' or visited[i][j]:
                continue

            q = deque()
            q.append((i, j))
            visited[i][j] = 1

            day = 0

            while q:
                x, y = q.popleft()
                day += int(maps[x][y])

                for dx, dy in d:
                    nx, ny = x + dx, y + dy
                    # 범위 내에 있고, 바다가 아니며, 방문하지 않은 경우
                    if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] != 'X' and not visited[nx][ny]:
                        visited[nx][ny] = 1
                        q.append((nx, ny))

            # 최대 며칠 동안 머무를 수 있는지를 리스트에 추가
            if day:
                answer.append(day)

    # 최종적으로 answer 리스트 반환
    if answer:
        answer.sort()
        return answer
    else:
        return [-1]

File: SooBin/240101_week/programmers_154540_무인도 여행.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;

int arr[100][100];
bool visited[100][100];

int dy[4] = {0, 0, 1, -1};
int dx[4] = {1, -1, 0, 0};

int M, N;

queue<pair<int, int>> q;

int bfs(int i, int j) {
    q.push({i, j});
    visited[i][j] = true;
    int cost = arr[i][j];

    while (!q.empty()) {
        int y = q.front().first;
        int x = q.front().second;
        q.pop();

        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny >= 0 && nx >= 0 && ny < M && nx < N && arr[ny][nx] != 0 && !visited[ny][nx]) {
                visited[ny][nx] = true;
                cost += arr[ny][nx];
                q.push({ny, nx});
            }
        }
    }
    return cost;
}

vector<int> solution(vector<string> maps) {
    vector<int> answer;
    M = maps.size();
    N = maps[0].size();

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            arr[i][j] = (maps[i][j] == 'X') ? 0 : maps[i][j] - '0';
            visited[i][j] = false;
        }
    }

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            if (arr[i][j] != 0 && !visited[i][j]) {
                answer.push_back(bfs(i, j));
            }
        }
    }

    if (answer.empty()) {
        answer.push_back(-1);
    } else {
        sort(answer.begin(), answer.end());
    }

    return answer;
}